clippy_lints/src/and_then_then_some.rs

```diff
@@ -0,0 +1,164 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet_with_applicability;
+use clippy_utils::{fn_def_id, match_def_path};
+use rustc_errors::Applicability;
+use rustc_hir::def::Res;
+use rustc_hir::{Block, Closure, Expr, ExprKind, FnDecl, HirId, Node, Pat, Path, QPath, Ty};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Detects usage of `Option::and_then` and `bool::then_some` that could
+    /// be replaced with `Option::filter`.
+    ///
+    /// ### Why is this bad?
+    /// Needless complexity, uses recent and uncommon stdlib functions instead of
+    /// one older function.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let x = Some("foo".to_string());
+    /// let _y = x.clone().and_then(|v| v.starts_with('f').then_some(v));
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let x = Some("foo".to_string());
+    /// let _y = x.clone().filter(|v| v.starts_with('f'));
+    /// ```
+    #[clippy::version = "1.81.0"]
+    pub AND_THEN_THEN_SOME,
+    nursery,
+    "detects usage of and_then and then_some that can be replaced by filter"
+}
+
+// note: `Option::filter` is older than `bool::then_some`,
+// so no msrv check is required.
+declare_lint_pass!(AndThenThenSome => [AND_THEN_THEN_SOME]);
+
+impl<'tcx> LateLintPass<'tcx> for AndThenThenSome {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        match expr.kind {
+            ExprKind::MethodCall(_, selfarg, [arg], _) | ExprKind::Call(_, [selfarg, arg]) => {
+                // TODO: check if type of reciever is diagnostic item Option?
```

Comment:
Yeah this should check if the type is `Option` then check the name of the call

Reply:
we're comparing the path later on, and the only way it could be a method call of one of Option's methods is if the receiver is Option, or if it's a function call that will fail typecheck, so checking the reciever type feels redundant.

originally i wanted it to still work even if the method was imported under a different name, but it doesn't seem like you can do that anyways.

Reply:
It's not redundant. The reason why paths are discouraged more often than not is because they aren't as stable as diagnostic items, reexports are essentially stripped after the resolver so if `Option` is moved to, say, `std::option_private::Option` and reexported in `Option`, it'll no longer work.

Reply:
Diagnostic item check + method name is the preferred way to do this